middleware.go

```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
+}
+
+func (l loggingInterceptor) intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp interface{}, err error) {
+	startTime := time.Now()
+	ctx = ctxlogrus.ToContext(ctx, logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l loggingInterceptor) interceptStream(
+	srv interface{},
+	ss grpc.ServerStream,
+	info *grpc.StreamServerInfo,
+	handler grpc.StreamHandler,
+) (err error) {
+	startTime := time.Now()
+	ctx := ctxlogrus.ToContext(ss.Context(), logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	wrapped := grpc_middleware.WrapServerStream(ss)
+	wrapped.WrappedContext = ctx
+
+	err = handler(srv, wrapped)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l *loggingInterceptor) requestFromContext(ctx context.Context, request *GRPCRequest) *GRPCRequest {
+	if d, ok := ctx.Deadline(); ok {
+		request.Deadline = d.UTC().Format(time.RFC3339Nano)
+	}
 
-			log.WithFields(fields).Info("Completed request")
+	if p, ok := peer.FromContext(ctx); ok && p != nil {
+		u := &url.URL{
+			Scheme: p.Addr.Network(),
+			Host:   p.Addr.String(),
+		}
+		request.PeerAddr = u.String()
+	}
+
+	if md, ok := metadata.FromIncomingContext(ctx); ok && md != nil {
+		request.UserAgent = strings.Join(md.Get("user-agent"), "")
+
+		if trace := md.Get("X-Cloud-Trace-Context"); len(trace) == 1 {
+			ctxlogrus.AddFields(ctx, logrus.Fields{"trace": trace[0]})
+		}
+	}
+
+	ctxlogrus.AddFields(ctx, logrus.Fields{"grpcRequest": request})
+
+	return request
+}
+
+// logStatus adds the gRPC Status to the log context.
+// If the response is an internal server error, log that as an Error
+// returns true if the logging was handled (e.g. internal server error)
+func (l *loggingInterceptor) log(ctx context.Context, err error, method string) {
+	if !l.filterRPC(ctx, method, err) {
+		return
+	}
+
+	if err != nil {
+		st := status.Convert(err)
+		jsonStatus, merr := ProtoJSONMarshalOptions.Marshal(st.Proto())
+		if merr != nil {
+			ctxlogrus.Extract(ctx).WithError(merr).Errorf("error marshalling error status into log")
+		}
+		ctxlogrus.AddFields(ctx, logrus.Fields{
+			"grpcStatus": json.RawMessage(jsonStatus),
 		})
+
+		// if the error was already logged by our custom error handler, we
+		// don't need to log the rest of this request
+		if handled := l.errorHandler(ctx, err, method); handled {
+			return
+		}
+
+		if st.Code() == codes.Internal {
+			ctxlogrus.Extract(ctx).WithError(err).Errorf("internal error response on RPC %s", method)
+			return
+		}
 	}
+
+	ctxlogrus.Extract(ctx).Infof("served RPC %v", method)
```

Comment:
it's super hard for me to grok when a giant chunk of the code is in an if statement. can you put this in an if err == nil  before all this?  then you can outdent the rest of this and it's simpler to see what's going on.

Reply:
I can add a comment in front of the error block.

I still want to log this even if error is not nil.
I'll rearrange it and add comments so it reads more clearer, but my purpose and intention here is to first add the gRPC Status to the log entry.
Only if the error that would be returned to the client is an internal error do I want it logging an Error.

Further I later added a custom error handler that I ended up not using in our services consuming this log handler, because it occurred to me we have the errors middleware that already converts everything to gRPC status without any of the additional flags and error context we add through the lifecycle of a request. This is where I had meant to do custom error handling, and we can still do that later replacing the errors middleware if we're so inclined.
But anyway the custom error handler is given the capability of logging for itself, returning truthy, and skipping the info-level logging (because it's been 'handled' already)

tl;dr: I'll try a different approach for parseability and lmk if that works or if you've got other ideas in the hat given what I'm trying to achieve here?

Reply:
ok ignore all that rambling above, I improved it by moving that block into `handleError` that itself calls the custom error handler so now it's just an if handled then return